### PR TITLE
[WIP] contrib/kube-prometheus: Add Thanos store StatefulSet

### PR DESCRIPTION
**Pull request by metalmatze**
The current Thanos extension misses the store as StatefulSet.

This PR is currently WIP as we need to discuss the way we want to handle secrets for GCS and S3 configuration-wise in kube-prometheus.

### CODE REVIEW

**Comment by brancz** [comment]
looks like this needs resolving some conflicts

File: contrib/kube-prometheus/jsonnet/kube-prometheus/kube-prometheus-thanos.libsonnet [hunk 1]
```diff
@@ -57,8 +57,56 @@ local servicePort = k.core.v1.service.mixin.spec.portsType;
...
+      local statefulSet = k.apps.v1beta2.statefulSet;
+      local volume = statefulSet.mixin.spec.template.spec.volumesType;
+      local container = statefulSet.mixin.spec.template.spec.containersType;
+      local containerEnv = container.envType;
+      local containerVolumeMount = container.volumeMountsType;
+
+      local labels = { app: 'thanos', 'thanos-peer': 'true' };
+
+      local c =
+        container.new('thanos-store', 'improbable/thanos:v0.1.0') +
```
Comment (elisiano) [inline contrib/kube-prometheus/jsonnet/kube-prometheus/kube-prometheus-thanos.libsonnet]: should the container image be `$._config.imageRepos.thanos + ':' + $._config.versions.thanos` instead? (like the `thanosQueryContainer` for example)

File: contrib/kube-prometheus/jsonnet/kube-prometheus/kube-prometheus-thanos.libsonnet [hunk 2]
```diff
@@ -57,8 +57,56 @@ local servicePort = k.core.v1.service.mixin.spec.portsType;
...
+        container.withArgs([
+          'store',
+          '--log.level=debug',
+          '--data-dir=/var/thanos/store',
+          '--cluster.peers=thanos-peers.' + $._config.namespace + '.svc.cluster.local:10900',
```
Comment (elisiano) [inline contrib/kube-prometheus/jsonnet/kube-prometheus/kube-prometheus-thanos.libsonnet]: I personally wouldn't use the `.cluster.local` suffix. Some users might have customized that.

**Comment by elisiano** [comment]
This implementation forces users to have s3 key/secret. 
Some users might actually be using `kube2iam` and use a role annotation so that the pod has access to the bucket without needing any credentials (https://gist.github.com/elisiano/a0c8cf83788f08bac577f25bcf830d10#file-kube-prometheus-thanos-compact-libsonnet-L41-L46 ). I was wondering if we could merge the logic in order to support both scenarios.

**Comment by brancz** [comment]
@elisiano I think your request is reasonable, we should integrate it.

**Comment by metalmatze** [comment]
Yep. This PR is work in progress and was merely a copy from what I've quickly written to get the things up in my cluster. It also leaves out GCS buckets entirely.

**Comment by metalmatze** [comment]
I'm adding the Thanos store StatefulSet with the update to v0.2 in #2275.

**Pull request closed without merging**